### ISSUE TITLE
csrf token for API calls without user session 

### ISSUE THREAD

**Issue by jardakotesovec** [pkp/pkp-lib#10311]
**Describe the issue**
When [accepting invitation](https://github.com/pkp/pkp-lib/issues/9658), UI needs to communicate with API, which requires csrf token, which is not present because it gets included only if the user is logged in, in the currentUser object:

```
window.pkp.currentUser = {
    "csrfToken": "y9Xqdi0JZJidbXGEg0Zfsmc8QVeHYdMEEcz4raIj",
    "id": 1,
    "roles": [
        1,
        16
    ]
}
```

**Suggested solution**
My understanding is that even if the user is not logged in - session gets created. So should be easy to keep the csrf there even without user being logged in?

So we could just move it elsewhere. But maybe keep it in original location as well, as its widely used.

New location could be:
```
window.pkp.context = {
{
    "csrfToken": "y9Xqdi0JZJidbXGEg0Zfsmc8QVeHYdMEEcz4raIj",
    "apiBaseUrl": "http://localhost:7003/index.php/publicknowledge/api/v1/",
    "pageBaseUrl": "http://localhost:7003/index.php/publicknowledge/en/",
    "legacyGridBaseUrl": "http://localhost:7003/index.php/publicknowledge/$$$call$$$/component/action"
}

```
As we already have api related urls there.

**What application are you using?**
 OJS, OMP or OPS version 3.5

#### Stable-3.5.0
[pkp-lib](https://github.com/pkp/pkp-lib) --> https://github.com/pkp/pkp-lib/pull/11602
[ojs](https://github.com/pkp/ojs) --> https://github.com/pkp/ojs/pull/4973 [TEST ONLY]

**Comment by jardakotesovec** [comment]
@touhidurabir @ewhanson @asmecher  Any thoughts on this one?

Once its triaged, I think it deserves "Try Me" label.

**Comment by ewhanson** [comment]
@kaitlinnewson, it is related but not the same issue. I talked with @defstat a bit about it last week. This is a case where there is neither a logged in user nor an API key.

@jardakotesovec, this approach make sense from an "accepting an invitation using the UI" perspective, but if the API endpoint requires neither a logged in user nor an API key, I think we would still run into trouble when accessing the API endpoint directly. In my conversation with @defstat, it felt like the "token" that comes along with the invitation is what would validate it as legitimate (in a similar way to password reset tokens) where there won't necessarily be an CSRF token available. 

To me it sounds like 1) this change would be good in general as I don't see why anonymous sessions shouldn't also have CSRF tokens, and 2) this solution would solve the use case described above, but not the underlying issue that a non `GET` API request requires a CSRF token.

**Comment by touhidurabir** [comment]
> the underlying issue that a non `GET` API request requires a CSRF token

I think that any api request does not need to be verified through CSRF token at all and only depend API Key/Token . But our use case if bit different as an user may not any api key associated with or more even , the app may not have `api secret` configured at all to have each user to have an api key. One solution can be is the force entry or auto generation of `API Secret` at installation time and then, on each user create/register, auto generation of API KEY . But then what about the change to api secret and corresponding impact of updating all available api key and the impact that follows it .  There is a lot to figure out and probably need a separate issue which may lead to much bigger API level infrastructural change.

@ewhanson  Now for the Invitation issue, how about we modify the `ValidateCsrfToken` to allow some `except` list which will not be validated for CSRF token . Any api routes can have middleware attached, so we can have a middleware for invitation related routes that will validate the invitation token attached to it and validated through it . That can be a solution for this specific case or similar cases where there is no logged in user or API KEY . 

@jardakotesovec I think what you have suggested will solve the specific use case described above and it's fine, but if we can solve it without depending on CSRF token , that will be better .

**Comment by ewhanson** [comment]
Thanks @touhidurabir, that sounds like a really good solution for the invitation use case. I think I'll spin that off as a separate issue to keep the conversation and work separate from this specific piece of work.

**Comment by asmecher** [comment]
I found [a good discussion](https://security.stackexchange.com/questions/166724/should-i-use-csrf-protection-on-rest-api-endpoints) of the relationship between CSRF and REST APIs:

> No cookies = No CSRF
>
> It really is that simple. Browsers send cookies along with all requests. CSRF attacks depend upon this behavior. If you do not use cookies, and don't rely on cookies for authentication, then there is absolutely no room for CSRF attacks, and no reason to put in CSRF protection.

So CSRF checks should be required for session-based access to API endpoints, but not for API key based access.

Just a note on our current API key support: until we improve our API capabilities (some older ideas at https://github.com/pkp/pkp-lib/discussions/8185), I consider API adoption risky. Each account can only have one API key, and that API key is equivalent to having login credentials. Those practices should both be improved -- I think @Vitaliy-1's work on moving users over to Eloquent and adding support for [spatie laravel-permissions](https://spatie.be/docs/laravel-permission/v6/introduction) will give us a big lift towards this.

**Comment by defstat** [comment]
In the case of Invitations API, as @ewhanson wrote above, there are 2 "problems":
1. There are requests that are not initiated by an existing user.
2. That also reflects on the global CSRF rule that is applied to all PUT, POST, DELETE requests.

There are 2 global middlewares that deal with that: `DecodeApiTokenWithValidation` and `ValidateCsrfToken`. 

I would expect that, if the `hasUser` authorizer is not present, those 2 middleware should not be applied.

**Comment by touhidurabir** [comment]
> In the case of Invitations API, as @ewhanson wrote above, there are 2 "problems":
> 
> 1. There are requests that are not initiated by an existing user.
> 2. That also reflects on the global CSRF rule that is applied to all PUT, POST, DELETE requests.
> 
> There are 2 global middlewares that deal with that: `DecodeApiTokenWithValidation` and `ValidateCsrfToken`.
> 
> I would expect that, if the `hasUser` authorizer is not present, those 2 middleware should not be applied.

There are 2 ways I can think about solving it 
1. we need to figure out what middleware attached to target route . However in the global middleware, we do get the `Router` but not the target `Route` to run as that will not be initiated before finalising the global middleware .  But we can figure out in the global middleware as such 
```php
$currentRoute = \PKP\core\PKPBaseController::getRequestedRoute($request);
$routeMiddleware = $currentRoute?->middleware() ?? [];

if (!empty($routeMiddleware) && !(in_array('has.user', $routeMiddleware) || in_array(HasUser::class, $routeMiddleware))) {
    // as `has.user` not defined for this route, we can skip it 
    return $next($request);
}
```
we can put this check in both the `DecodeApiTokenWithValidation` and `ValidateCsrfToken` middleware . 

2. Another option is to have a `except` property for both the `DecodeApiTokenWithValidation` and `ValidateCsrfToken` and check those accordingly like https://github.com/laravel/framework/blob/11.x/src/Illuminate/Foundation/Http/Middleware/VerifyCsrfToken.php#L85

**Comment by defstat** [comment]
@taslangraham @touhidurabir @jardakotesovec is this issue, perhaps, resolved by #10279?

**Comment by asmecher** [comment]
Moving off to 3.5.0-1 -- but echoing @defstat's question above... @taslangraham @touhidurabir @jardakotesovec

**Comment by asmecher** [comment]
@touhidurabir, please don't forget this one!

**Comment by touhidurabir** [comment]
> > Moving off to 3.5.0-1 -- but echoing [@defstat](https://github.com/defstat)'s question above... [@taslangraham](https://github.com/taslangraham) [@touhidurabir](https://github.com/touhidurabir) [@jardakotesovec](https://github.com/jardakotesovec)
> 
> No, [#10279](https://github.com/pkp/pkp-lib/issues/10279) did not resolve this issue. The two are slightly related, but not the same. [#10279](https://github.com/pkp/pkp-lib/issues/10279) fixed an issue where the middleware was checking for API Tokens in query params while tokens were being passed via Authorization header. The `ValidateCsrfToken` middleware is designed to skip CSRF validation whenever there's a token in the request.
> 
> However, for this case there is neither a logged in user nor an API Token, therefore the CSRF check gets executed and fails.
> 
> [#10342](https://github.com/pkp/pkp-lib/issues/10342) seemed to have provided a work around. [@touhidurabir](https://github.com/touhidurabir) could you confirm if it does?

@asmecher @jardakotesovec @taslangraham @defstat [#10342](https://github.com/pkp/pkp-lib/issues/10342) does provide an alternative solution for the invitation's API from where this issue originated at first . However it does not provide exact solution for what describe in the issue description . 

But as the original case resolved by [#10342](https://github.com/pkp/pkp-lib/issues/10342) , I am not sure if this need right now anymore . I think it can be closed as `not planned` for now and we can reopen and work on this if necessity rise in future .

**Comment by jardakotesovec** [comment]
Right, so whats currently happening is that when accepting invitation - its not sending real csrf token, because its not available in pkp.currentUser.csrfToken if user is not logged in.

So I guess the "fix" is that the API does not require it for not logged in users?
I don't think its big deal in this case - because its already behind invitation key.

But I see that we do create session even if the user is not logged in and we do send csrf token for example when user is logging in. So I don't see reason why we would not support csrf token for unlogged in user?

**Comment by touhidurabir** [comment]
> So I don't see reason why we would not support csrf token for unlogged in user?

@jardakotesovec you are right and there is no harm adding it by default even for non logged in users . In fact laravel also does that (see https://laravel.com/docs/11.x/csrf) and problem will only happen in case of `XSS` vulnerability which is a different case to handle . 

I see that `currentUser.csrfToken` is being used many places and moving it will be a breaking change at this point (which also mentioned in the issue description) . And after a bit discussion with @asmecher , we think it's better to follow laravel's guideline to have this as `meta` in header . 

However, I see that in for non logged mode, `window.pkp` is undefined so having it in `window.pkp.context` will not have it for non logged in user. so what I am missing here ?

**Comment by touhidurabir** [comment]
@asmecher can you review the PRs at https://github.com/pkp/pkp-lib/issues/10311#issue-2465432575 .

**Comment by jardakotesovec** [comment]
Having in meta header I think is fine as well. So I am in support of PR above.

**Comment by asmecher** [comment]
Merged and cherry-picked forward to `main`.
